phoenix-core-client/src/main/java/org/apache/phoenix/expression/util/bson/DocumentComparisonExpressionUtils.java

```diff
@@ -0,0 +1,370 @@
+          Preconditions.checkArgument(bsonValue instanceof BsonArray,
+              "Expected type for Bson value is Array for $or operator");
+          BsonArray bsonArray = (BsonArray) bsonValue;
+          List<BsonValue> bsonValues = bsonArray.getValues();
+          for (BsonValue value : bsonValues) {
+            if (evaluateExpression(document, (BsonDocument) value)) {
+              return true;
+            }
+          }
+          return false;
+        }
+        case "$and": {
+          BsonValue bsonValue = conditionExpression.get(firstFieldKey);
+          Preconditions.checkArgument(bsonValue instanceof BsonArray,
+              "Expected type for Bson value is Array for $and operator");
+          BsonArray bsonArray = (BsonArray) bsonValue;
+          List<BsonValue> bsonValues = bsonArray.getValues();
+          for (BsonValue value : bsonValues) {
+            if (!evaluateExpression(document, (BsonDocument) value)) {
+              return false;
+            }
+          }
+          return true;
+        }
+        default: {
+          throw new IllegalArgumentException(firstFieldKey + " is not a known operator");
+        }
+      }
+    }
+  }
+
+  private static boolean isExists(final BsonDocument document,
+      final BsonDocument conditionExpression) {
+    Set<Map.Entry<String, BsonValue>> entrySet = conditionExpression.entrySet();
+    Preconditions.checkArgument(entrySet.size() == 1, "Expected entry for the exists operation"
+        + " is 1");
+    for (Map.Entry<String, BsonValue> bsonValueEntry : entrySet) {
+      String fieldKey = bsonValueEntry.getKey();
+      BsonValue bsonValue = bsonValueEntry.getValue();
+      Preconditions.checkArgument(bsonValue instanceof BsonDocument,
+          "Expected type for Bson value is Document for exists operation");
+      BsonDocument bsonDocument = (BsonDocument) bsonValue;
+      BsonValue existsValue = bsonDocument.get(EXISTS_OP);
+      Preconditions.checkArgument(existsValue instanceof BsonBoolean,
+          "Expected type for $exists value is boolean");
+      BsonBoolean existsValBoolean = (BsonBoolean) existsValue;
+      if (existsValBoolean.getValue()) {
+        return exists(fieldKey, document);
+      } else {
+        return !exists(fieldKey, document);
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Returns true if the field provided in the condition expression has value greater than
+   * or equals to the value provided in the condition expression.
+   * Condition Expression format:
+   * {
+   *   <field>: {
+   *     "$gte": <value>
+   *   }
+   * }
+   *
+   * @param document The document used for comparison.
+   * @param conditionExpression Condition Expression Document.
+   * @return True if the field provided in the condition expression has value greater than
+   * or equals to the value provided in the condition expression.
+   */
+  private static boolean greaterThanOrEquals(final BsonDocument document,
```

Comment:
Can we combine all these comparison methods into one by passing the comparison operator instead of having a separate method for each operator?

Reply:
Yes i can make this, but since we need separate methods in SQL style comparison, i kept it this way:

- SQL style expression -> contains individual comparison methods -> CommonComparisonExpressionUtils -> single method for all comparisons
- Document style expression -> contains individual comparison methods -> CommonComparisonExpressionUtils -> single method for all comparisons

When i introduce ANTLR style expression evaluation, i can make this change as well:

- SQL style expression -> single method for all comparisons -> CommonComparisonExpressionUtils -> single method for all comparisons
- Document style expression -> single method for all comparisons -> CommonComparisonExpressionUtils -> single method for all comparisons

@kadirozde does this sound good?

Reply:
You can still have a separate method for each of them if you need but they should call a common method to eliminate duplicate code.

Reply:
I am suggesting something like this:
```
private static boolean compare(final BsonDocument document,
      final BsonDocument conditionExpression, CommonComparisonExpressionUtils.CompareOp op) {
    Set<Map.Entry<String, BsonValue>> entrySet = conditionExpression.entrySet();
    Preconditions.checkArgument(entrySet.size() == 1, "Expected entry size for a compare operation"
        + " is 1");
    for (Map.Entry<String, BsonValue> bsonValueEntry : entrySet) {
      String fieldKey = bsonValueEntry.getKey();
      BsonValue bsonValue = bsonValueEntry.getValue();
      Preconditions.checkArgument(bsonValue instanceof BsonDocument,
          "Expected type for Bson value is Document for a compare operation");
      BsonDocument bsonDocument = (BsonDocument) bsonValue;
      BsonValue compareValue = bsonDocument.get(getBsonCompareOp(op)); // need to implement getBsonCompareOp
      BsonValue topLevelValue = document.get(fieldKey);
      BsonValue actualValue = topLevelValue != null ? topLevelValue
          : CommonComparisonExpressionUtils.getFieldFromDocument(fieldKey, document);
      return actualValue != null && CommonComparisonExpressionUtils.compareValues(actualValue,
          compareValue, op);
    }
    return false;
  }
private static boolean greaterThanOrEquals(final BsonDocument document,
      final BsonDocument conditionExpression) {

    return compare(document, conditionExpression, CommonComparisonExpressionUtils.CompareOp.GREATER_OR_EQUAL);
}
private static boolean greaterThan(final BsonDocument document,
      final BsonDocument conditionExpression) {

    return compare(document, conditionExpression, CommonComparisonExpressionUtils.CompareOp.GREATER);
}
and so on
```

Reply:
I see, you mean within the individual classes SQLComparisonExpressionUtils and DocumentComparisonExpressionUtils also, we should have common comparison method, which can directly call common comparison method from CommonComparisonExpressionUtils.

Sounds good improvement, let me change it now.

Reply:
Done